### PR TITLE
ignore to propagate service account token

### PR DESCRIPTION
**Pull request by Garrybest**
Signed-off-by: Garrybest <garrybest@foxmail.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Do not propagate service account token because it should be created by the member cluster.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
I guess you mean `Secret` with `kubernetes.io/service-account-token` type should prevent from propagating?

**Comment by Garrybest** [comment]
Right. It is created by service account.

File: pkg/detector/detector.go [hunk 1]
```diff
@@ -223,6 +223,10 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
 		return err
 	}
 
+	if shouldIgnorePolicyMatching(object) {
```
Comment (XiShanYongYe-Chang) [inline pkg/detector/detector.go:226]: Would it be more appropriate to put this logic into `func (d *ResourceDetector) EventFilter?

Comment (Garrybest) [reply pkg/detector/detector.go:226]: Good idea!

Comment (Garrybest) [reply pkg/detector/detector.go:226]: Well, I just realized that we need to get the origin object. However, the parameter of `EventFilter` is `ClusterWideKey`.  So we need to retrieve the origin object in `EventFilter`. Maybe it's not a proper solution......

**Comment by pigletfly** [comment]
I don't know if there others cases that should be ignored like Secret with kubernetes.io/service-account-token case?Or make this a option(not hardcoded), so users can add a case if they find one and needn't modify karmada code.

**Comment by RainbowMango** [comment]
Before starting to think of the solution, I want to know the use case further. So, can you provide more details about how to reproduce it? And what's the harm with it?

**Comment by Garrybest** [comment]
OK, I will show some phenomenon later.

**Comment by Garrybest** [comment]
First create a service account.

```bash
$ kubectl create ns test
$ kubectl create sa -n test sa-test
$ kubectl get secrets -n test 
NAME                  TYPE                                  DATA   AGE
sa-test-token-vgk99   kubernetes.io/service-account-token   3      6s
```

If we have a policy to automatically propagate all secret in the namespace into member clusters like this.

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: secret-propagation
  namespace: test
spec:
  resourceSelectors:
    - apiVersion: v1
      kind: Secret
      namespace: test
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
        - member3
```

The secret in member cluster will be created by karmada-agent and deleted by kube-controller-manager. So the secret list will be like:

```
$ kubectl --context member3 get secrets -n test -w
NAME                  TYPE                                  DATA   AGE
default-token-22vbx   kubernetes.io/service-account-token   3      13m
sa-test-token-b6dwd   kubernetes.io/service-account-token   3      3m48s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      1s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      1s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
sa-test-token-jcrx8   kubernetes.io/service-account-token   3      0s
```

And the logs of karmada-agent will be full of recreating.

```
$ kubectl --context member3 logs -n karmada-system karmada-agent-556c77b744-pphpr --tail 20 
I0325 12:54:54.576068       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:54.576231       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:54.673445       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:54.677178       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:54.677531       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:54.770444       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:54.774362       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:54.774577       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:54.874182       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:54.877874       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:54.878313       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:54.976196       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:54.980284       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:54.980423       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:55.074174       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:55.081445       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:55.081661       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
I0325 12:54:55.174619       1 workstatus_controller.go:269] recreating cluster=member3, v1, kind=Secret, test/sa-test-token-jcrx8
I0325 12:54:55.178262       1 objectwatcher.go:103] Created resource(kind=Secret, test/sa-test-token-jcrx8) on cluster: member3
I0325 12:54:55.178820       1 workstatus_controller.go:226] reflecting Secret(test/sa-test-token-jcrx8) status to Work(karmada-es-member3/sa-test-token-jcrx8-5b4b5f67d4)
```

**Comment by Garrybest** [comment]
This kind of secret should never be propagated into member clusters for sake of protecting karmada-agent.

**Comment by Garrybest** [comment]
> I don't know if there others cases that should be ignored like Secret with kubernetes.io/service-account-token case?Or make this a option(not hardcoded), so users can add a case if they find one and needn't modify karmada code.

Maybe it's hard. Taken this PR by example, the field `type` in `Secret` is not common in any other kind of objects.

**Comment by RainbowMango** [comment]
> The secret in member cluster will be created by karmada-agent and deleted by kube-controller-manager.

Why `kube-controller-manager` deletes the secret? 

Does the `ServiceAccount` (named `sa-test`) has been propagated?

**Comment by Garrybest** [comment]
The secret in member cluster has an annotation: 

```
kubernetes.io/service-account.uid: d6e2dc25-8c83-4ad3-bac6-dc97c08ee832
```

This uid belongs to the ServiceAccount in karmada controll-plane.

```bash
$ kubectl get sa -n test sa-test -oyaml                                    
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: "2022-03-25T12:41:47Z"
  name: sa-test
  namespace: test
  resourceVersion: "3120"
  uid: d6e2dc25-8c83-4ad3-bac6-dc97c08ee832
secrets:
- name: sa-test-token-jcrx8
```

No matter whether we propagate the ServiceAccount, the uid will not exist in member cluster, so `kube-controller-manager` tries to delete the secret.

**Comment by RainbowMango** [comment]
cc @mrlihanbo @XiShanYongYe-Chang 
Please take a look.

**Comment by RainbowMango** [comment]
Link to the previous PRs.
- https://github.com/karmada-io/karmada/pull/655#issuecomment-910050932
- https://github.com/karmada-io/karmada/pull/660

**Review by XiShanYongYe-Chang — COMMENTED** [review]
lgtm

**Comment by RainbowMango** [comment]
Sorry, I missed this. I'll be back on this after the holiday.

/assign

**Comment by RainbowMango** [comment]
Let me summarize this issue:

The `Secret`, with type `kubernetes.io/service-account-token`,  which is created along with `ServiceAccount` should be prevented from propagating. Because:
- There is no use case to propagate the `Secret` as the member cluster will create it automatically.
- If it is propagated to member clusters, the `kube-controller-manager` will delete it as a garbage resource.

**Comment by RainbowMango** [comment]
I think we'd better ignore this kind of resource from [EventFilter](https://github.com/karmada-io/karmada/blob/756fd12dd303980282e921b21d1b3212ef66c133/pkg/detector/detector.go#L286), rather than put it in reconcile queue. 
How do you think?

**Comment by XiShanYongYe-Chang** [comment]
> I think we'd better ignore this kind of resource from [EventFilter](https://github.com/karmada-io/karmada/blob/756fd12dd303980282e921b21d1b3212ef66c133/pkg/detector/detector.go#L286), rather than put it in reconcile queue. How do you think?

I agree with this idea, the current writing method is a bit fragmented, but it is not feasible based on the current input parameter of EventFilter.

https://github.com/karmada-io/karmada/pull/1525#discussion_r833131779

**Comment by RainbowMango** [comment]
Right, let me take a second look.

File: pkg/detector/detector.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/detector/detector.go:226]: The parmater of `EventFilter` is `not` `ClusterWideKey`, it's the object. And we don't need to retrieve the original object from apiserver or cache.

Comment (RainbowMango) [reply pkg/detector/detector.go:226]: If I understand right, this piece of code should work.
```diff
diff --git a/pkg/detector/detector.go b/pkg/detector/detector.go
index 4f0da7ea..51e355e1 100644
--- a/pkg/detector/detector.go
+++ b/pkg/detector/detector.go
@@ -318,6 +318,15 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
                return false
        }

+       // The secret, with type 'kubernetes.io/service-account-token', is created along with `ServiceAccount` should be
+       // prevented from propagating.
+       // Refer to https://github.com/karmada-io/karmada/pull/1525#issuecomment-1091030659 for more details.
+       if st, ok := obj.(corev1.Secret); ok {
+               if st.Type == corev1.SecretTypeServiceAccountToken {
+                       return false
+               }
+       }
+
        return true
 }

```

Comment (Garrybest) [reply pkg/detector/detector.go:226]: Great, that's true. I made a mistake.

Comment (RainbowMango) [reply pkg/detector/detector.go]: > Maybe because we use a dynamic informer manager.

Yes, exactly. I forgot this.

**Review by RainbowMango — COMMENTED** [review]
LGTM
cc @XiShanYongYe-Chang Please take a look.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**